service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/ConditionLinkSetMapper.java

```diff
@@ -0,0 +1,155 @@
+    private static final String RELATED_CLINICAL_CONTENT_URL = "https://fhir.hl7.org.uk/STU3/StructureDefinition/Extension-CareConnect-RelatedClinicalContent-1";
+    private static final String ACTIVE = "active";
+    private static final String MAJOR = "major";
+    private static final String LIST = "List";
+    private static final String EXTENSION_NOT_PRESENT = "Extension not present";
+
+    private final MessageContext messageContext;
+
+    public String mapConditionToLinkSet(Condition condition, boolean isNested) {
+        var builder = ConditionLinkSetMapperParameters.builder()
+            .isNested(isNested)
+            .linkSetId(messageContext.getIdMapper().getOrNew(ResourceType.Condition, condition.getIdElement().getIdPart()));
+
+        buildEffectiveTimeLow(condition).map(builder::effectiveTimeLow);
+        buildEffectiveTimeHigh(condition).map(builder::effectiveTimeHigh);
+        buildAvailabilityTime(condition).map(builder::availabilityTime);
+        builder.relatedClinicalContent(buildRelatedClinicalContent(condition));
+
+        var qualifier = buildQualifier(condition);
+        qualifier.map(value -> value.equalsIgnoreCase(MAJOR)).map(builder::qualifierIsMajor);
+        qualifier.map(builder::qualifier);
+
+        var clinicalStatus = buildClinicalStatusCode(condition);
+        clinicalStatus.map(value -> value.equalsIgnoreCase(ACTIVE)).map(builder::clinicalStatusIsActive);
+        clinicalStatus.map(builder::clinicalStatusCode);
+
+        var conditionNamed = buildConditionNamed(condition);
+        if (conditionNamed.isPresent()) {
+            conditionNamed
+                .filter(value -> !value.equalsIgnoreCase(EXTENSION_NOT_PRESENT))
+                .map(builder::conditionNamed);
+        } else {
+            String newId = messageContext.getIdMapper().getNew();
+            builder.generateObservationStatement(true);
+            builder.conditionNamed(newId);
+            buildPertinentInfo(condition).map(builder::pertinentInfo);
+        }
+
+        return TemplateUtils.fillTemplate(OBSERVATION_STATEMENT_TEMPLATE, builder.build());
+    }
+
+    private Optional<String> buildConditionNamed(Condition condition) {
+        Optional<Extension> actualProblemExtension = ExtensionMappingUtils.filterExtensionByUrl(condition, ACTUAL_PROBLEM_URL);
+        if (actualProblemExtension.isPresent()) {
+            Optional<Reference> reference = actualProblemExtension
+                .map(Extension::getValue)
+                .map(value -> (Reference) value);
+
+
+            if (reference.map(this::checkIfReferenceIsObservation).orElse(false)) {
+                return reference
+                    .map(ref -> messageContext.getIdMapper().getOrNew(ref));
+            }
+            return Optional.empty();
+        } else {
+            return Optional.of(EXTENSION_NOT_PRESENT);
+        }
```

Comment:
It should be possible to do something much simpler here. This fails a test though.

```suggestion
            return ExtensionMappingUtils.filterExtensionByUrl(condition, ACTUAL_PROBLEM_URL)
            .map(Extension::getValue)
            .map(value -> (Reference) value)
            .filter(this::checkIfReferenceIsObservation)
            .map(ref -> messageContext.getIdMapper().getOrNew(ref));
```